### DISCUSSION TITLE
Areas for improvement

### DISCUSSION THREAD

**Discussion by danzuep** [danzuep/MailKitSimplified#67] [Ideas]
Things that could be improved in the design.

**Comment by danzuep** [comment]
Implement proper builder patterns.
The current implementation is more fluent than builder.

**Comment by danzuep** [comment]
Have a shared base/core/domain layer.

**Comment by danzuep** [comment]
Make it generic so libraries other than MailKit can be used.

**Comment by danzuep** [comment]
Add benchmark tests.

**Comment by mabeale** [comment]
Hi, great library!
Perhaps something that is missing is being able to listen to flag changes on emails.

If I set up my monitor as such:
`await _imapReceiverClient.MonitorFolder
    .SetMessageSummaryItems()
    .SetIgnoreExistingMailOnConnect()
    .OnMessageArrival(OnArrivalAsync)
    .IdleAsync(_cancel.Token);`
    
There doesn't appear to be an extension method or handler for the `IMailFolder.MessageFlagsChanged` event.

Are there any plans to add this or have I missed something obvious?

**Reply by danzuep** [reply]
Thanks Mark, great idea! Is this something you are looking to use?

**Reply by danzuep** [reply]
I've added an separate thread for this excellent idea.

**Reply by mabeale** [reply]
Excellent! Yes, when I start my idle process I am ignoring any read emails in the inbox but I would like to process them once they become unread.
I then want to move the processed email to a sub folder, which I've seen can prove tricky if the ImapClient is busy in another thread, i.e. the idle process.

But for now, I just need the IMonitorFolder to have a method for `OnMessageFlagsChanged` if possible.

Thanks a lot!
